build_variables.bzl

```diff
@@ -744,6 +744,9 @@ libtorch_cuda_distributed_base_sources = [
 libtorch_cuda_distributed_extra_sources = [
     "torch/csrc/distributed/c10d/NCCLUtils.cpp",
     "torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp",
+    "torch/csrc/distributed/c10d/ProcessGroupUCC.cpp",
```

Comment:
See: https://github.com/pytorch/pytorch/pull/79918/files#r906266685

Reply:
Thanks for raising this issue! I got to understand your concern now. I agree that as it stands, UCC PG is not compiled when USE_CUDA=0. 
Nevertheless, for code readability I think it is simpler to have UCC implementation in one place  rather than breaking them into CPU-only and GPU pieces.
Can we instead solve the problem with the compiler configuration? 
For example, I have tried adding UCC files (e.g., ProcessGroupUCC.cpp, UCCUtils.cpp) to both libtorch_distributed_base_sources and libtorch_cuda_distributed_extra_sources.
Doing this seems to allow me to run UCC PG in both CPU-mode (compiling with USE_CUDA=0) and in GPU-mode (compiling with USE_CUDA=1). Let me know if you have other suggestions. Thanks.

Reply:
I don't understand the build system of PyTorch very well, so I have no idea about what will happen if you do that. I am just wondering if you add them to both `libtorch_distributed_base_sources` and `libtorch_cuda_distributed_extra_sources`, is this effectively equal to copy-pasting the code twice? How does the linker know which to choose when encountering the same symbol?

Reply:
I have a theory why it works, but I need to confirm on the compiler/linker workflow as Pytorch is kinda complex with CMake, Bazel etc.
In any case, I feel perhaps a more robust way is to perform explicit check on CUDA (e.g., with Bazel's if_cuda), and include the UCC files to the right file glob accordingly. I'd be happy as long as we can solve the problem with compiler config so that the UCC files can be kept clean.
As you mentioned, let's keep this PR as is and handle this issue later (I will try get advice from a Pytorch build expert).

Reply:
@zasdfgbnm linker should fail with conflicting symbols. Or, if none of the functions are marked as `TORCH_API` it will link it it twice and use twice the memory